server/src/main/java/org/elasticsearch/common/util/CollectionUtils.java

```diff
@@ -100,44 +99,68 @@ public static int[] toArray(Collection<Integer> ints) {
      * @param messageHint A string to be included in the exception message if the call fails, to provide
      *                    more context to the handler of the exception
      */
-    public static void ensureNoSelfReferences(Object value, String messageHint) {
-        Iterable<?> it = convert(value);
-        if (it != null) {
-            ensureNoSelfReferences(it, value, Collections.newSetFromMap(new IdentityHashMap<>()), messageHint);
+    public static void ensureNoSelfReferences(final Object value, final String messageHint) {
+        ensureNoSelfReferences(value, Collections.newSetFromMap(new IdentityHashMap<>()), messageHint);
+    }
+
+    private static void ensureNoSelfReferences(final Object value, final Set<Object> ancestors, final String messageHint) {
```

Comment:
We definitely need comments here explaining why it's like this

Reply:
++, you're absolutely right.

Reply:
Done, a0317bb94941e31d6464921a3feff5a632e53d5c